tools/AutomationTestFramework/Management/RunbookProcessor.ps1

```diff
@@ -0,0 +1,207 @@
+    $sessionFunctions = Get-ChildItem function:
+    foreach ($testFile in $testFiles) {
+        # Get the function list from the script file
+        . "$path\$testFile"
+        # https://sqljana.wordpress.com/2015/09/23/perform-set-operations-union-intersection-minus-complement-using-powershell/
+        $scriptFunctions = Get-ChildItem function: | Where-Object { $sessionFunctions -inotcontains $_ }
+        $testFunctions = $scriptFunctions | Where-Object { 
+                $desc = (Get-Help $_).Description
+                $_.Name -ilike 'Test*' -and $desc -ne $null -and $desc[0].Text -contains $tag
+            }
+        $scriptFunctions | ForEach-Object { Remove-Item function:\$_ }
+       
+        if ($testFunctions.Count -eq 0) {
+            "# $testFile`: No tests found"
+            continue
+        }
+
+        $null = $variableList.Add($testFile.BaseName)
+        "# $testFile"
+        "`$$($testFile.BaseName) = @("
+        for ($i = 0; $i -lt $testFunctions.Count; $i++) {
+            $testSuffix = if($i -eq $testFunctions.Count - 1) { ' )' } else { ',' }
+            "`t'$($testFunctions[$i])'$testSuffix"
+        }
+    }
+
+    $testListSuffix = if($variableList.Count -eq 0) { ' @()' } else { '' }
+    "`$testList =$testListSuffix"
+    for ($i = 0; $i -lt $variableList.Count; $i++) {
+        $varSuffix = if($i -eq $variableList.Count - 1) { '' } else { ' +' }
+        "`t`$$($variableList[$i])$varSuffix"
+    }
+}
+
+function Create-Runbooks (
+    [hashtable] $template,
+    [string] $srcPath,
+    [string[]] $projectList,
+    [string] $outputPath) {
+
+    if (-not (Test-Path $outputPath)) {
+        $null = New-Item -ItemType directory -Path $outputPath -ErrorAction Stop
+    } else { 
+        Write-Verbose "Cleaning up the $outputPath folder..."
+        Remove-Item "$outputPath\*" -ErrorAction Stop
+    }
+    Write-Verbose "Collecting .ps1 test files..."
+    foreach($folder in Get-TestFolders $srcPath $projectList) {
+        $bookName = "Live$($folder.Name)Tests"
+        $bookPath = "$outputPath\$bookName.ps1"
+        $null = New-Item $bookPath -type file -Force
+        $loginParamsTemplate = '%LOGIN-PARAMS%'
+        $testListTemplate = '%TEST-LIST%'
+        Get-Content $template.Path | ForEach-Object {
+            $content = switch -wildcard ($_) {
+                "*$loginParamsTemplate" {
+                    $_ -replace $loginParamsTemplate, "'$($template.AutomationConnectionName)' '$($template.SubscriptionName)'"
+                } $testListTemplate {
+                    Create-TestContent $folder.Path | Out-String
+                } default {
+                    $_
+                }
+            }
+            $content | Add-Content $bookPath
+        }
+        Write-Verbose "$bookPath generated."
+    }
+}
+
+# https://www.codeisahighway.com/how-to-easily-and-silently-obtain-accesstoken-bearer-from-an-existing-azure-powershell-session/
+function Get-AzureRmAccessToken() {
```

Comment:
So, why do you need an access token?  You should be able to use Get-AzureRmContext to get the current context, and AzureSession.ClientFactory to create any management clients you need.

Reply:
This is used to log in as the current Azure account in the Job's app domain:
```
        $accessToken = Get-AzureRmAccessToken
        $accountId = $(Get-AzureRmContext).Account.Id
        $subscriptionName = $(Get-AzureRmContext).Subscription.Name
        Write-Verbose "Starting '$bookName' runbook..."
        $null = Start-Job `
            -Name $bookName `
            -ArgumentList $accessToken,$accountId,$subscriptionName,$bookName,$automation `
            -ScriptBlock {
            param($accessToken,$accountId,$subscriptionName,$bookName,$automation)
            Login-AzureRmAccount -AccessToken $accessToken -AccountId $accountId -Subscription $subscriptionName
            Start-AzureRmAutomationRunbook -Name $bookName -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -ErrorAction Stop -Wait -MaxWaitSeconds 3600
        }
```

I looked for a **while** to find a solution for this. There is no documentation on logging in as the currently logged in person (in a separate session). I looked over the parameter sets for `Add-AzureRmAccount` and found that the access token method seemed to make the most sense. So, as per the link in the comment, someone on the internet found a way to get the access token from the current user. Just read that post and you'll understand what I was searching for, and the challenges of doing this.

If this is an oversight in our documentation, we should look into addressing it.